app/scripts/controllers/metamask-notifications/metamask-notifications.ts

```diff
@@ -0,0 +1,954 @@
+   */
+  async #getUserStorage(): Promise<UserStorage | null> {
+    const userStorageString: string | null =
+      await this.#storage.getNotificationStorage();
+
+    if (!userStorageString) {
+      return null;
+    }
+
+    try {
+      const userStorage: UserStorage = JSON.parse(userStorageString);
+      return userStorage;
+    } catch (error) {
+      log.error('Unable to parse User Storage');
+      return null;
+    }
+  }
+
+  /**
+   * @deprecated - This needs rework for it to be feasible. Currently this is a half-baked solution, as it fails once we add new triggers (introspection for filters is difficult).
+   *
+   * Checks for the complete presence of trigger types by group across all addresses in user storage.
+   *
+   * This method retrieves the user storage and uses `MetamaskNotificationsUtils` to verify if all expected trigger types for each group are present for every address.
+   * @returns A record indicating whether all expected trigger types for each group are present for every address.
+   * @throws {Error} If user storage does not exist.
+   */
+  public async checkTriggersPresenceByGroup(): Promise<
+    Record<TRIGGER_TYPES_GROUPS, boolean>
+  > {
+    const userStorage = await this.#getUserStorage();
+    this.#assertUserStorage(userStorage);
+
+    // Use MetamaskNotificationsUtils to check the presence of triggers
+    return MetamaskNotificationsUtils.checkTriggersPresenceByGroup(userStorage);
+  }
+
+  /**
+   * Verifies the presence of specified accounts and their chains in user storage.
+   *
+   * This method retrieves the user storage and uses `MetamaskNotificationsUtils` to check if the specified accounts and all their supported chains are present in the user storage.
+   *
+   * @param accounts - An array of account addresses to be checked for presence.
+   * @returns A record where each key is an account address and each value is a boolean indicating whether the account and all its supported chains are present in the user storage.
+   * @throws {Error} If user storage does not exist.
+   */
+  public async checkAccountsPresence(
+    accounts: string[],
+  ): Promise<Record<string, boolean>> {
+    // Retrieve user storage
+    const userStorage = await this.#getUserStorage();
+    this.#assertUserStorage(userStorage);
+
+    // Use MetamaskNotificationsUtils to check the presence of accounts
+    return MetamaskNotificationsUtils.checkAccountsPresence(
+      userStorage,
+      accounts,
+    );
+  }
+
+  /**
+   * Toggles the enabled state of metamask notifications.
+   *
+   * This method checks for the presence of a BearerToken token and a storage key before attempting to toggle the notification state.
+   * If either the BearerToken token or the storage key is missing, the method disables metamask notifications and logs an error.
+   * Otherwise, it toggles the current state of metamask notifications (enabled/disabled).
+   *
+   * @async
+   * @throws {Error} If updating the state fails.
+   */
```

Comment:
If notifications are toggled off, and the user adds new accounts, and then notifications are toggled on, do we need to do something like `updateOnChainTriggersByAccount` for those accounts?

Reply:
When notifications are toggled on we'll call `createOnChainTriggers()` which will go through all accounts and subscribe them to notifications - so it should cover any new accounts that were added.

We can probably add some smarter logic in the future - e.g. saving users preferences to have accounts that were turned off previously stay turned off.

Reply:
Note this is def hard for an external reviewer to review - since this controller is kind-of decoupled from the UI so it is hard to get some context.

I'll make sure that these public methods are well documented and give an idea of where they are used.

Reply:
I've updated the JSDoc comments here to include an `**Action**` on the intent of each method.